openoa/utils/downloader.py

```diff
@@ -0,0 +1,428 @@
+    except Exception as e:
+        logger.error('Failed to make connection to cds: '+ str(e))
+        logger.error('Please see: https://cds.climate.copernicus.eu/api-how-to')
+        raise NameError(e)
+
+    # create save_pathname if it does not exist
+    save_pathname = Path(save_pathname).resolve()
+    if not save_pathname.exists():
+        save_pathname.mkdir()
+
+    # downloads all years from today back to the year 2000
+    now = datetime.datetime.now()
+    years = list(range(2000,now.year+1,1))
+
+    # get the data for the closest 9 nodes to the coordinates
+    node_spacing = 0.250500001*1
+
+    # download the data
+    for year in years:
+
+        outfile = save_pathname / f"{save_filename}_{year}.nc"
+
+        if year == now.year:
+            months = list(range(1,now.month,1))
+        else:
+            months = list(range(1,12+1,1))
+
+        # See: https://cds.climate.copernicus.eu/cdsapp#!/dataset/reanalysis-era5-single-levels-monthly-means?tab=form
+        # for formulating other requests from cds
+        if not outfile.is_file() or year==now.year:
+
+            logger.info(f"Downloading ERA5: {outfile}")
+
+            try:
+                c.retrieve(
+                    "reanalysis-era5-single-levels-monthly-means",
+                    {
+                        "product_type": "reanalysis",
+                        "format": "netcdf",
+                        "variable": [
+                            "10m_wind_speed", "2m_temperature", "surface_pressure",
+                        ],
+                        "year": year,
+                        "month": months,
+                        "product_type": "monthly_averaged_reanalysis",
+                        "time": [
+                            "00:00"
+                        ],
+                        "area": [
+                            lat+node_spacing, lon-node_spacing, 
+                            lat-node_spacing, lon+node_spacing,
+                        ],
+                    },
+                    outfile)
+            except:
+                logger.error(f"Failed to download ERA5: {outfile}")
+
+
+    # get the saved data
+    ds_nc = xr.open_mfdataset(f"{save_pathname / f'{save_filename}*.nc'}")
+
+    # rename variables to conform with OpenOA
+    ds_nc = ds_nc.rename_vars({"si10":"windspeed_ms","t2m":"temperature_K","sp":"surf_pres_Pa"})
+    
+    # select the central node only for now
+    if 'expver' in ds_nc.dims:
+        sel = ds_nc.sel(expver=1,latitude=lat,longitude=lon, method="nearest")
+    else:
+        sel = ds_nc.sel(latitude=lat,longitude=lon, method="nearest")   
+
+    # convert to a pandas dataframe
+    df = sel.to_dataframe()
+
+    # select required columns
+    df = df[["windspeed_ms","temperature_K","surf_pres_Pa"]]
+
+    # rename the index to match other datasets
+    df.index.name = "datetime"
+
+    # drop any empty rows
+    df = df.dropna()
+
+    if save_filename is not None:
```

Comment:
I would suggest explaining in the docstring how passing a value of None for "save_filename" means only a data frame gets returned instead of saving a csv file.

Reply:
Set to always save the file, as actually the save_filename is required for other parts of the function